pkg/controller/ttl/ttlcontroller.go

```diff
@@ -0,0 +1,221 @@
+	if !cache.WaitForCacheSync(stopCh, ttlc.hasSynced) {
+		return
+	}
+
+	workers := 5
+	for i := 0; i < workers; i++ {
+		go wait.Until(ttlc.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down TTL controller")
+}
+
+func (ttlc *TTLController) addNode(obj interface{}) {
+	node, ok := obj.(*v1.Node)
+	if !ok {
+		glog.Errorf("Unexpected object types: %v", obj)
+		return
+	}
+	atomic.AddInt32(&ttlc.nodeCount, 1)
+	ttlc.enqueueNode(node)
+}
+
+func (ttlc *TTLController) updateNode(_, newObj interface{}) {
+	node, ok := newObj.(*v1.Node)
+	if !ok {
+		glog.Errorf("Unexpected object type: %v", newObj)
+		return
+	}
+	ttlc.enqueueNode(node)
+}
+
+func (ttlc *TTLController) deleteNode(obj interface{}) {
+	node, ok := obj.(*v1.Node)
+	if !ok {
+		glog.Errorf("Unexpected object types: %v", obj)
+		return
+	}
+	atomic.AddInt32(&ttlc.nodeCount, -1)
+	ttlc.enqueueNode(node)
+}
+
+func (ttlc *TTLController) enqueueNode(node *v1.Node) {
+	key, err := controller.KeyFunc(node)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v", node)
+		return
+	}
+	ttlc.queue.Add(key)
+}
+
+func (ttlc *TTLController) worker() {
+	workFunc := func() bool {
+		key, quit := ttlc.queue.Get()
+		if quit {
+			return true
+		}
+		defer ttlc.queue.Done(key)
+
+		err := ttlc.updateNodeIfNeeded(key.(string))
+		if err == nil {
+			ttlc.queue.Forget(key)
+			return false
+		}
+
+		ttlc.queue.AddRateLimited(key)
+		utilruntime.HandleError(err)
+		return false
+	}
+	for {
+		if quit := workFunc(); quit {
+			glog.Infof("TTL controller worker shutting down")
+			return
+		}
+	}
+}
+
+// TODO: Tune it a bit more.
+func (ttlc *TTLController) getDesiredTTL() int {
+	nodeCount := atomic.LoadInt32(&ttlc.nodeCount)
+	switch {
+	case nodeCount <= 100:
+		return 0
+	case nodeCount <= 1000:
+		return 30
+	case nodeCount <= 2000:
+		return 60
+	default:
+		return 300
+	}
+}
+
+func getIntFromAnnotation(node *v1.Node, annotationKey string) (int, bool) {
+	if node.Annotations == nil {
+		return 0, false
+	}
+	annotationValue, ok := node.Annotations[annotationKey]
+	if !ok {
+		return 0, false
+	}
+	intValue, err := strconv.Atoi(annotationValue)
+	if err != nil {
+		glog.Warningf("Cannot convert the value %q with annotation key %q for the node %q",
+			annotationValue, annotationKey, node.Name)
+		return 0, false
+	}
+	return intValue, true
+}
+
+func setIntAnnotation(node *v1.Node, annotationKey string, value int) {
+	if node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+	node.Annotations[annotationKey] = strconv.Itoa(value)
+}
+
+func (ttlc *TTLController) updateNodeIfNeeded(key string) error {
+	obj, exists, err := ttlc.nodeStore.GetByKey(key)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		glog.Infof("Node %s no longer exists", key)
+		return nil
+	}
+	node := obj.(*v1.Node)
+	desiredTTL := ttlc.getDesiredTTL()
+
+	if value, ok := getIntFromAnnotation(node, ObjectTTLAnnotation); !ok || value != desiredTTL {
```

Comment:
(optional) nit: return early to avoid more indentation.

```go
value, ok := getIntFromAnnotation(node, ObjectTTLAnnotation); 
if ok && value == desired {
  return nil
}
```